src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more than readable.
+                    //"this" is only readable in members of readonly structs, unless we are in a constructor.
+                    //"this" is not returnable by reference in a struct.
+                    if (!thisref.Type.IsValueType ||
+                        (thisref.Type.IsReadOnly && (this.ContainingMember() as MethodSymbol)?.MethodKind != MethodKind.Constructor) ||
```

Comment:
Yes. it should be. I assumed that we can assign to initonly fields in methods nested in constructors. It is not actually the case (as it should be, for lambdas in any case). Will change and add tests. Good catch!

---
In reply to: [118784631](https://github.com/dotnet/roslyn/pull/19713#discussion_r118784631) [](ancestors = 118784631)